ipq40xx: Support for ZyXEL WSQ50

jasongaunt opened the pull request:
ipq40xx: add support for the ZyXEL WSQ50

### Introduction

This patch adds support for the ZyXEL WSQ50 device:

(Taken from [https://deviwiki.com/wiki/ZyXEL_WSQ50](https://deviwiki.com/wiki/ZyXEL_WSQ50))

- SOC:  IPQ4019 / QCA Dakota
- CPU:  Quad-Core ARMv7 Processor rev 5 (v7l) Cortex-A7
- DRAM: 512 MiB DDR3L
- NOR:  8 Mb Micron M25P80 Serial Flash Embedded Memory
- ETH: Qualcomm Atheros QCA8075 Gigabit Switch (3 x LAN, 1 x WAN)
- USB: 1 x 2.0 (via Synopsys DesignWare DWC3 controller in the SoC)
- WLAN1: Qualcomm Atheros QCA4018 2.4GHz 802.11bgn 2:2x2
- WLAN2: Qualcomm Atheros QCA4018 5GHz 802.11a/n/ac 2:2x2
- WLAN3: Qualcomm Atheros QCA9984 5GHz 802.11a/n/ac Wave2 4:4x4
- BT: Qualcomm CSR8811 BT 4.1 LE
- INPUT: RESET Button
- LEDS: Singular RGB LED broken out into 3 channels (red / green / blue) 

All of the above hardware is working with this patch. 

### Initial installation

This device needs a 3.3v logic USB serial adapter to flash the initial firmware, as well as some knowledge of u-boot and a custom ZyXEL firmware unlocking script. You can install via USB or over Ethernet. 

If you want to install the firmware from a USB stick, make sure this formatted as FAT32, you've copied across the `*-factory.bin` image to it and placed it in the USB port on the back of the unit.

#### Prepare the device

1. Unplug all cables, pull off the rubber feet, remove 4 screws and then carefully lever the device case apart
2. Locate the UART header on the front side of the PCB (by the LYNwave.com stamp) - it's an unsoldered 0.1" pitch 4 pin header at the front of the unit (opposite the Ethernet ports)
3. Connect a suitable serial adapter (soldering on pins if you need to), the pinouts are as follows: `[ V | G |  RX | TX ]` (3.3v, Ground, RX and TX respectively) - V is the square copper trace on the PCB (closest the USB port), the rest are usual round copper traces
4. Open the serial adapter with the usual `115200 8N1` settings in your favourite terminal application
5. Reset the unit back to factory defaults - hold the RESET button whilst attaching power, you'll eventually see `Erase rootfs_data partition...` over the UART
6. Power cycle the unit again and when prompted, press any key to stop autoboot
7. Follow the instructions on this web page to unlock the bootloader: [ZyXEL U-Boot Unlocker](https://akao.co.uk/tools/zyxel_unlocker/)
8. With the bootloader now unlocked, run `run boot_mmc_1` to boot into the failsafe firmware
9. When booting you should be able to log in as `root` / `1234` (but you may need to enter failsafe mode and `mount_root` and `passwd` to change it)

#### Installing via USB

1. Locate where the USB mounted, typically under `/tmp/storage/`, this may help you easily find it: `cd /tmp/storage/*/*`
2. Use the native firmware flasher `FW_upgrade` (note the capitalisation, there's another lowercase `fw_upgrade` which this wraps) to flash the firmware: `FW_upgrade ./openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.bin`

#### Installing via Ethernet

1. Plug in an Ethernet cable to the WAN port connected your home LAN and then run `udhcpc` to get an IP
2. Activate DNS It won't have DNS, so run `echo "nameserver 8.8.8.8" > /etc/resolv.conf`
3. Download the firmware from your favourite webserver with `curl -k --output /tmp/newfw.bin https://example.com/openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.bin`
4. Use the native firmware flasher `FW_upgrade` (note the capitalisation, there's another lowercase `fw_upgrade` which this wraps) to flash the firmware: `FW_upgrade /tmp/newfw.bin`

When either of those have finished, `reboot` the unit and enjoy OpenWrt properly :)

Pictures and full documentation will come with a wiki page when approved, however the process is as follows;

### Future upgrades

The usual sysupgrade process is supported, just install updates via the OpenWrt LuCI GUI.

### Booting the failsafe firmware again

If for whatever reason you want to boot ZyXEL's failsafe firmware again (God help you), just repeat steps 1-8 above and go through the usual setup procedure and then install any firmware update for it to overwrite OpenWrt.

Signed-off-by: Jason Gaunt <github@akao.co.uk>

robimarko commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -48,7 +48,8 @@ ALLWIFIBOARDS:= \
 	zte_mf287plus \
-	zyxel_nbg7815
+	zyxel_nbg7815 \
+	zyxel_wsq50
```
Is the BDF for this board already included in the https://github.com/openwrt/firmware_qca-wireless repo?

jasongaunt replied:
Hi robimarko, no it is not and I don't know how to rectify that

robimarko replied:
Make a PR with the correct BDF-s wrapped by https://github.com/qca/qca-swiss-army-knife/blob/master/tools/scripts/ath10k/ath10k-bdencoder

jasongaunt replied:
Thank you for assisting on IRC with this Robert, PR raised here; https://github.com/openwrt/firmware_qca-wireless/pull/21

jasongaunt replied:
This is now merged in

robimarko commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -29,6 +29,9 @@ ipq40xx_setup_interfaces()
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"
 		;;
+	zyxel,wsq50)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" "wan"
```
There are already boards using the same config, just add yours there, there is no need to duplicate code.

jasongaunt replied:
Will do 👍

jasongaunt replied:
This has been fixed - extra code removed and used another config

robimarko commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -69,21 +69,6 @@ askey_do_upgrade() {
 	nand_do_upgrade "$1"
 }
 
-zyxel_do_upgrade() {
```
This needs to be a separate commit

robimarko replied:
Still needs to be done

robimarko commented on target/linux/ipq40xx/base-files/lib/upgrade/zyxel.sh:
```diff
@@ -0,0 +1,107 @@
+
+		case "$rootfs" in
+			"/dev/mmcblk0p5")
+				kernel="/dev/mmcblk0p4"
```
Please no hardcoded partitions

jasongaunt replied:
This was based off the similar script in https://github.com/openwrt/openwrt/blob/main/target/linux/ipq806x/base-files/lib/upgrade/zyxel.sh with hard-coded partitions - I opted to do this differently to allow rolling back to stock firmware, but it still needs to know where it's installed - this is a specificity of the partition layout of the WSQ50

robimarko replied:
But, they have labels in the partition table, dont they?

So, why not find the partition by the label instead of hardcoding them?

robimarko replied:
Cant you use `find_mmc_part` here?

jasongaunt replied:
Good shout! Swapping to that now :)

jasongaunt replied:
This has been fixed - Swapped to `find_mmc_part` like you suggested, still needed to use the initial paths in the case statement though

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+	};
+
+	soc {
+		rng@22000 {
```
With the exception of `tcsr` nodes all of these have existing labels, so please just use those instead of this.
Example:
https://github.com/openwrt/openwrt/blob/74e7f8ebbdc19c58ac59c792154041a6b3c124f5/target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4029-ws-ap391x.dts#L122

jasongaunt replied:
This has been fixed - Followed your advice and moved them all to labels, all still working perfectly :)

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+			function = "blsp_uart1";
+			bias-disable;
+		};
+	};
```
Newline after

jasongaunt replied:
This has been fixed - Added newline here and also cleaned up rest of DTS to match

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+
+&tlmm {
+	serial1_pins: serial1_pinmux {
+		mux {
```
Drop these mux subnodes

jasongaunt replied:
I believe these are required because the pins are different than in the DTSI

robimarko replied:
Pins are, but you can put them directly under `serial1_pinmux`, there is no need to have them in the `mux` subnode.

jasongaunt replied:
This has been fixed - mux nodes removed

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+			output-high;
+			bias-pull-up;
+			drive-strength = <0x08>;
+			pins = "gpio44", "gpio45", "gpio46";
```
Please pins first, function second.

jasongaunt replied:
This has been fixed - pins put to top in whole DTS where relevant

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+		mux {
+			output-high;
+			bias-pull-up;
+			drive-strength = <0x08>;
```
No hex numbers here, 8 is fine enough.

jasongaunt replied:
This has been fixed - All hex numbers converted to integers where possible

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+	bluetooth {
+		compatible = "csr,8811";
+
+		enable-gpios = <&tlmm 32 GPIO_ACTIVE_HIGH>;
```
Is this actually used in code, like is anything using this?

Cause if so, then the hog above is completely redundant and this would be race condition as to who claims this resource first.

jasongaunt replied:
I can possibly clean this up

jasongaunt replied:
This has been fixed - Code was no longer required after identifying bluetooth module lived on USB 3.0 bus instead on this particular device

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+};
+
+&swport1 {
+	status = "disabled";
```
No need to have this node at all, they are disabled by default.

jasongaunt replied:
Yep that's why I put in the "okay" to turn them back on again? They were present on the original factory firmware

robimarko replied:
This one is disabled in your DTS, effectively doing nothing.

jasongaunt replied:
This has been fixed - Stray node removed

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+			qcom,ath10k-calibration-variant = "Zyxel-WSQ50";
+			qcom,cal-len = <0x2f20>;
+			qcom,cal-offset = <0x9000>;
+			qcom,mtd-name = "0:ART";
```
These look like copy/paste from stock DTS, these are not used upstream.

So, drop all of the `qcom,*` properties except for the variant.

jasongaunt replied:
This has been fixed - Stray stanzas removed, device works perfectly still as firmware is loaded in by hotplug.d script to correct addresses

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -1267,3 +1274,22 @@ define Device/zyxel_wre6606
+	DEVICE_ALT0_MODEL := WSQ60
+	SOC := qcom-ipq4019
+	KERNEL_SIZE := 4096k
+	BLOCKSIZE := 64k
```
Are you sure?

That sounds really low for eMMC

jasongaunt replied:
The BLOCKSIZE? Yes
The KERNEL_SIZE? Unsure, that was stolen from elsewhere

robimarko replied:
Then you can drop `KERNEL_SIZE` as its unused

jasongaunt replied:
This has been fixed - `KERNEL_SIZE` removed as robimarko was correct, it's superfluous

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -1267,3 +1274,22 @@ define Device/zyxel_wre6606
+		ath10k-firmware-qca4019-ct ath10k-firmware-qca9984-ct \
+		kmod-fs-ext4 kmod-mmc kmod-fs-squashfs tune2fs e2fsprogs \
+		losetup blockd kmod-ath3k kmod-bluetooth uboot-envtools
+	$(call Device/ZyXELImageLzma)
```
This goes first

jasongaunt replied:
This has been fixed - Code cleaned up

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -1267,3 +1274,22 @@ define Device/zyxel_wre6606
+	RAS_ROOTFS_SIZE := 20934k
+	RAS_VERSION := "$(VERSION_DIST) $(REVISION)"
+	DEVICE_PACKAGES := \
+		ath10k-firmware-qca4019-ct ath10k-firmware-qca9984-ct \
```
ath10k-firmware-qca4019-ct is included by default

jasongaunt replied:
This has been fixed - Superfluous package removed

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -1267,3 +1274,22 @@ define Device/zyxel_wre6606
+	RAS_VERSION := "$(VERSION_DIST) $(REVISION)"
+	DEVICE_PACKAGES := \
+		ath10k-firmware-qca4019-ct ath10k-firmware-qca9984-ct \
+		kmod-fs-ext4 kmod-mmc kmod-fs-squashfs tune2fs e2fsprogs \
```
Remove `kmod-mmc` and `kmod-fs-squashfs` these are already compiled in the kernel

jasongaunt replied:
This has been fixed - As before, superfluous packages removed

robimarko commented on target/linux/ipq40xx/image/generic.mk:
```diff
@@ -1267,3 +1274,22 @@ define Device/zyxel_wre6606
+	DEVICE_PACKAGES := \
+		ath10k-firmware-qca4019-ct ath10k-firmware-qca9984-ct \
+		kmod-fs-ext4 kmod-mmc kmod-fs-squashfs tune2fs e2fsprogs \
+		losetup blockd kmod-ath3k kmod-bluetooth uboot-envtools
```
Why do you need blockd?

Remove `uboot-envtools` as its included by default

jasongaunt replied:
blockd was added so that we can make use of 2 additional unused partitions in flash out of the box:

* mmcblk0p9 - label `bu1` - 256 MiB
* mmcblk0p10 - label `bu2` - 3.0 GiB

 (this came at the request of another device owner as well)

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+	status = "okay";
+
+	tricolor: lp5562@30 {
+		clock-mode = [01];
```
Replace all of these decompiled copy/paste values from the stock DTS with proper values..

Also, move this below the cells.

Check the DT bindings of the driver for what is proper format:
https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/Documentation/devicetree/bindings/leds/leds-lp55xx.yaml?h=next-20231122

jasongaunt replied:
This has been fixed - thanks @robimarko that was quite insightful, have migrated to that and it still works as expected

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+&blsp1_spi1 {
+	status = "okay";
+
+	m25p80@0 {
```
flash@0

jasongaunt replied:
This has been fixed - Renamed node

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+	m25p80@0 {
+		compatible = "jedec,spi-nor", "n25q128a11";
+		spi-max-frequency = <24000000>;
+		reg = <0x00>;
```
Convert to decimal

jasongaunt replied:
This has been fixed - Converted to decimal

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+			partition2@160000 {
+				label = "env";
+				reg = <0x00160000 0x00010000>;
+				read-only;
```
I assume this is U-Boot env so why is it read only?

jasongaunt replied:
The label matches what is actually in the chip, the factory ROM kept it as read-only but I guess there's no reason to keep it as such.

robimarko replied:
I am sorry, but `env` on its own is rather vague.

Is it the U-boot environment or?

jasongaunt replied:
Yep it's the u-boot environment, I didn't know if u-boot itself depended on this, I'll try renaming it and see what happens

jasongaunt replied:
Tried renaming this in the DTS file, however `cat /proc/mtd` still shows it as "env"

jasongaunt replied:
Marking this as resolved - there's very little I can do about this

robimarko replied:
Can you dump the live DTS?

I suspect you are hitting the same issue as other boards do, in so that the bootloader will patch in the nodes here.

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+
+		status_red: chan@0 {
+			chan-name = "red:status";
+			// linux,default-trigger = "default-on";
```
If this is not required, remove it, same goes for other instances in this node.

jasongaunt replied:
This has been fixed - Stray trigger stanzas removed

robimarko commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+		status_red: chan@0 {
+			chan-name = "red:status";
+			// linux,default-trigger = "default-on";
+			led-cur = [c0];
```
Same as for `clock-mode`, this is not valid.

jasongaunt replied:
This has been fixed - Moved to `/bits/ 8 <??>` format and also done the others

jasongaunt commented on target/linux/ipq40xx/files/arch/arm/boot/dts/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,316 @@
+	status = "okay";
+
+	m25p80@0 {
+		compatible = "jedec,spi-nor", "n25q128a11";
```
Unclear, it was in the factory decompiled DTS and spotted in similar devices in OpenWrt's codebase, I'll try without it

robimarko replied:
It is sometimes left as a bootloader depends on the `"n25q128a11"` compatible that QCA abuses to patch in the partition table

jasongaunt replied:
This has been fixed - It was not needed and has been removed :)

aioue reviewed (commented):
Install tested and working on factory shipped unit.

robimarko commented on target/linux/ipq40xx/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -49,6 +49,10 @@ case "$FIRMWARE" in
 		caldata_extract_mmc "0:ART" 0x9000 0x2f20
 		ath10k_patch_mac $(mmc_get_mac_binary ARTMTD 0x12)
 		;;
+	zyxel,wsq50)
```
I would suggest doing this as well as other caldata extractions directly via NVMEM, then you can also set the MAC via DTS as well instead of relying on userpace.

robimarko commented on target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh:
```diff
@@ -37,7 +37,8 @@ preinit_set_mac_address() {
 		;;
-	zyxel,nbg6617)
+	zyxel,nbg6617|\
+	zyxel,wsq50)
```
This should be done via NVMEM, then there is no need for this

jasongaunt replied:
Migrated to NVMEM and confirmed working

robimarko commented:
Can you rebase this, I will give it another round of reviews, sorry for the delay

arkdev1 commented:
@robimarko hi, it's possible to force the rebase?

robimarko commented:
I can force it, but its not my PR and code, so its up to the author if he is still interested

jasongaunt commented:
Sorry all, I have had some significant health issues lately and life issues to deal with. Rebasing and testing this now poses 2 challenges;

1.  I am struggling to find the time for other projects lately, let alone this
2. All of my WSQ50's are now back in the usual (hard to reach) places

That said, I don't want to let this die. @robimarko I need to pick up with you on IRC at some point, there's some occasional stability issues I don't quite know how to resolve. Please keep this open for now and we'll progress it when I can scrape together the time.

robimarko commented:
No worries, take your time, life is more important than these devices

syned42 reviewed (commented):
reviewed. works.

arkdev1 commented:
Hi @robimarko,
sorry about the tag.
I'm trying to install OpenWrt on the WSQ50 product, but I'm having problems with the firmware build:

```
/home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.15.134/scripts/dtc/dtc -O dtb -i/home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.15.134/arch/arm/boot/dts/ -Wno-unit_address_vs_reg -Wno-simple_bus_reg -Wno-unit_address_format -Wno-pci_bridge -Wno-pci_device_bus_num -Wno-pci_device_reg -Wno-avoid_unnecessary_addr_size -Wno-alias_paths -Wno-graph_child_address -Wno-graph_port -Wno-unique_unit_address    -o /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-nbg6617.dtb /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-nbg6617.dtb.tmp
rm -f /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4019-mf289f.dtb.tmp
cc1: fatal error: /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.15.134/arch/arm/boot/dts/-wsq60.dts: No such file or directory
compilation terminated.
/home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.15.134/scripts/dtc/dtc -O dtb -i/home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.15.134/arch/arm/boot/dts/ -Wno-unit_address_vs_reg -Wno-simple_bus_reg -Wno-unit_address_format -Wno-pci_bridge -Wno-pci_device_bus_num -Wno-pci_device_reg -Wno-avoid_unnecessary_addr_size -Wno-alias_paths -Wno-graph_child_address -Wno-graph_port -Wno-unique_unit_address    -o /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-wsq50.dtb /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-wsq50.dtb.tmp
make[5]: *** [Makefile:19: /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image--wsq60.dtb] Error 1
make[5]: *** Waiting for unfinished jobs....
rm -f /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-nbg6617.dtb.tmp
rm -f /home/arkdev1/zyxelopenwrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/image-qcom-ipq4018-wsq50.dtb.tmp
make[5]: Leaving directory '/home/arkdev1/zyxelopenwrt/target/linux/ipq40xx/image'
make[4]: *** [Makefile:23: install] Error 2
make[4]: Leaving directory '/home/arkdev1/zyxelopenwrt/target/linux/ipq40xx'
make[3]: *** [Makefile:11: install] Error 2
make[3]: Leaving directory '/home/arkdev1/zyxelopenwrt/target/linux'
time: target/linux/install#28.35#15.38#29.83
    ERROR: target/linux failed to build.
make[2]: *** [target/Makefile:30: target/linux/install] Error 1
make[2]: Leaving directory '/home/arkdev1/zyxelopenwrt'
make[1]: *** [target/Makefile:24: /home/arkdev1/zyxelopenwrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/stamp/.target_install] Error 2
make[1]: Leaving directory '/home/arkdev1/zyxelopenwrt'
make: *** [/home/arkdev1/zyxelopenwrt/include/toplevel.mk:232: world] Error 2

```

robimarko commented:
My guess is that you are missing `SOC :=` because the DTS name is wrong

jasongaunt commented:
Minor progress update, I'm trying to secure another WSQ50 from the usual auction sites so I can start dev work on this again. All my other units are in inaccessible places now so I'm getting a 4th unit to develop on.

arkdev1 commented:
> Minor progress update, I'm trying to secure another WSQ50 from the usual auction sites so I can start dev work on this again. All my other units are in inaccessible places now so I'm getting a 4th unit to develop on.

I bought two unit from Amazon. I'm from Italy, and it cost me around seventy euros each.
If i can help, i can send you one.
Fun fact: The Multy Application is broken, so I can't get it to work with the stock firmware.

jasongaunt commented:
> > Minor progress update, I'm trying to secure another WSQ50 from the usual auction sites so I can start dev work on this again. All my other units are in inaccessible places now so I'm getting a 4th unit to develop on.
> 
> I bought two unit from Amazon. I'm from Italy, and it cost me around seventy euros each. If i can help, i can send you one. Fun fact: The Multy Application is broken, so I can't get it to work with the stock firmware.

That's super kind of you to offer but I will buy my own 😊

And yeah, the whole reason I did this was ZyXEL make good hardware but not so much other things... 😁

If you're just looking to get the thing working again, attached you'll find the last build I did for all of the above.
Note that you will _not_ be able to use package manager or get sysupdates, it's a development build after all.
You will however be able to use this to install the _correct_ version through official channels once it's all done.
[openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.zip](https://github.com/openwrt/openwrt/files/14888554/openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.zip)

arkdev1 commented:
> > > Minor progress update, I'm trying to secure another WSQ50 from the usual auction sites so I can start dev work on this again. All my other units are in inaccessible places now so I'm getting a 4th unit to develop on.
> > 
> > 
> > I bought two unit from Amazon. I'm from Italy, and it cost me around seventy euros each. If i can help, i can send you one. Fun fact: The Multy Application is broken, so I can't get it to work with the stock firmware.
> 
> That's super kind of you to offer but I will buy my own 😊
> 
> And yeah, the whole reason I did this was ZyXEL make good hardware but not so much other things... 😁
> 
> If you're just looking to get the thing working again, attached you'll find the last build I did for all of the above. Note that you will _not_ be able to use package manager or get sysupdates, it's a development build after all. You will however be able to use this to install the _correct_ version through official channels once it's all done. [openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.zip](https://github.com/openwrt/openwrt/files/14888554/openwrt-ipq40xx-generic-zyxel_wsq50-squashfs-factory.zip)

Thank you very much! I will give it a try.

jasongaunt commented:
New WSQ50 secured, development for this is being picked back up again.

aioue commented:
Anything I can do to help @jasongaunt ?

jasongaunt committed:
ipq40xx: Separate ZyXEL sysupgrade code into dedicated script

Migrates ipq40xx ZyXEL sysupgrade script to separate file in prep for adding WSQ50

Signed-off-by: Jason Gaunt <github@akao.co.uk>

jasongaunt commented:
@aioue Nope! Have rebased and built a new image with the latest commit from main, testing it out now on one of my units :)

It came up okay, can connect to it etc, just stability testing now.

jasongaunt committed:
ipq40xx: add support for ZyXEL WSQ50

Adds support for ZyXEL WSQ50 - more info here; https://github.com/openwrt/openwrt/pull/14028

Signed-off-by: Jason Gaunt <github@akao.co.uk>

robimarko commented on package/boot/uboot-envtools/files/ipq40xx:
```diff
@@ -76,6 +76,9 @@ linksys,whw03v2)
 	ubootenv_add_uci_config "/dev/mtd6" "0x0" "0x10000" "0x10000"
 	;;
+zyxel,wsq50)
+	ubootenv_add_uci_config "/dev/mtd2" "0x0" "0x10000" "0x10000"
```
Instead of this, you can just add `compatible = "u-boot,env";` to the partition that is holding the env.

robimarko commented on target/linux/ipq40xx/base-files/lib/upgrade/platform.sh:
```diff
@@ -212,7 +212,8 @@ platform_do_upgrade() {
 		nand_do_upgrade "$1"
 		;;
-	zyxel,nbg6617)
+	zyxel,nbg6617 |\
```
Remove the whitespace after `nbg6617`

robimarko commented on target/linux/ipq40xx/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq4019-wsq50.dts:
```diff
@@ -0,0 +1,345 @@
+	};
+
+	chosen {
+		bootargs-append = " clk_ignore_unused fstools_ignore_partname=1";
```
Remove this

arkdev1 commented:
Hi @jasongaunt, any news, i'm triyng to use the router as a mesh point, with no success.

GoetzGoerisch commented:
Thank you for your PR.

Just a heads up, we had a recent [treewide change from ZyXEL to Zyxel](https://github.com/openwrt/openwrt/commit/3774f3272ee976b922306a881b6eb603e79f81ed).
Please adjust your commit message accordingly. Thank you!

syned42 commented:
Hello,
any news about a final firmware for WSQ50 ?
I use the snapshot firmware for over 1 year, but Im not able to update any parts of software.
Is there a chance to get WSQ50 to be part of offical branch?

aioue commented:
> Hello, any news about a final firmware for WSQ50 ? I use the snapshot firmware for over 1 year, but Im not able to update any parts of software. Is there a chance to get WSQ50 to be part of offical branch?

Development has stopped as the author has moved on to other hardware. The work to rebase do the zyxel find/replace would be an excellent starter issue, if you wished to take up the mantle.

DrDimes83 commented:
> > Hello, any news about a final firmware for WSQ50 ? I use the snapshot firmware for over 1 year, but Im not able to update any parts of software. Is there a chance to get WSQ50 to be part of offical branch?
> 
> Development has stopped as the author has moved on to other hardware. The work to rebase do the zyxel find/replace would be an excellent starter issue, if you wished to take up the mantle.

I Have this Router here and I don’t now how to start.  I Flasht Openwrt on it with FDTI. 
I think that this thing must be done fast. 
So, can anybody help me fix this thing and make it right Working Official?
It is so sad to see that the Author is not writing about his moving to other Hardware. I purchased this Hardware cause the Tripple Wifi and The 4Gigs of EMMC. 
So never mind, i need Help and a step by step Guide

greetings Dimes

syned42 commented:
Jasongaunt you said, you have the last Openwrt build installed. Is it again a snapshot or a final compiled 23.05 with posibility to install more software packages ? Could you store your build anywhere ?
My WSQ50 has only a a snapshot presented 2023 - I cant install any packages and no patches.

DrDimes83 commented:
> Hi all - I stopped working on this because of two reasons; 1. I needed WiFi 7 for VR and 2. I was struggling for time and couldn't keep up with the expected changes and in particular, rebase requests because "the goalposts moved". This isn't a dig at @robimarko - I know and trust he's only trying to promote _good_ code, but I found it quite challenging to keep on top of whilst maintaining all the other plates life asks us to keep spinning.
> 
> That said, I still have four WSQ50's here, all with the last OpenWrt build installed, two even have the debug UART exposed to breadboard-friendly jumper wires. I am happy to _donate_ these to a seasoned OpenWrt developer (such as robimarko) or as a last resort, try and scrounge some time together to rebuild against the latest release. The latter could take weeks / months.


Hmmm… on 24.10.xx they write Supporting over 1500 Devices. 
You have done 90% of the Work and the snapshot is fine, but without Packages the Openwrt is to basic and has no Opportunities as devices with so much possibilities and support. 
So if Openwrt is not working, is it possible to install the WSQ60 Firmware? 
I just try it and startet this with putty and putty is writing wrong Header after i start applying the WSQ60 Firmware.

After all, thank you for your time and Development. 

Greetings Dimes

DrDimes83 commented:
> Jasongaunt you said, you have the last Openwrt build installed. Is it again a snapshot or a final compiled 23.05 with posibility to install more software packages ? Could you store your build anywhere ? My WSQ50 has only a a snapshot presented 2023 - I cant install any packages and no patches.

It’s only the Snapshot Version, there is no official version. I testet everything out. The newest Version is on the Server from [https://wolfnet.akao.co.uk/tftp/](url) from Jun 2024

andreas5232 commented:
@aioue thanks for the fast response! Sounds like a tough learning curve... Then I'm sorry and have to pass the work to someone else.

GoetzGoerisch commented:
@aioue @andreas5232 do you have time to test and give your feedback?
@jasongaunt how to proceed from you point of view.

jasongaunt commented:
Hi all, I moved house late last year and took the opportunity to upgrade to 2.5 Gbe throughout, as such I'm no longer using these access points. 

I am struggling to find the time to work on this, I have too many projects on my plate. If anyone wants to take the reigns I'm happy to donate the two WSQ50's I have to a worthy developer. Both are slightly modded with embedded ESP32's that function as WiFi Telnet gateways to get direct UART access remotely.

GoetzGoerisch commented:
@jasongaunt If you have time, please reset this branch to the last commit of my branch and squash the last to commits, keeping my SoB.
Or I can open a new PR carrying over your work and get I merged?

Could you contribute and license the zyxel_unlocker as well please? e.g. add a License to the source code to copy and paste it as code to the OpenWrt wiki?

andreas5232 commented:
> We need feedback for running it on the proper hardware that it can be accepted into main. Therefore, you could try my build output or build yourself and image based on my uptodate branch. Then I would propose a new PR where this needs to be approved by users of the hardware and this is then merged if no other feedback is blocking this. Then it would run in the regualar build processes and uptodate images are downloadable from the firmware selector or downlods page.

I have now two WSQ50 lying on my desk and waiting desperately for openwrt. As far as my knowledge goes, I should be able to install the recent build using all the information provided in this thread. However I'm not sure if I'll also manage to build it on my own yet.

Do you have a recent build I can use for testing?
